db/backup/recovery_test.py

```diff
@@ -0,0 +1,246 @@
+            )
+            self.assertEqual(sample, prod_random_results[0])
+
+    def test_random_rows_family_participant(self):
+        """ Test 10 random rows in the family_participant table """
+        restored_random_results = get_results(
+            f'SELECT * FROM family_participant ORDER BY RAND() LIMIT 10',
+            self.local_connection,
+        )
+        for entry in restored_random_results:
+            family_id = entry[0]
+            participant_id = entry[1]
+
+            prod_random_results = get_results(
+                f"SELECT * FROM family_participant FOR SYSTEM_TIME AS OF TIMESTAMP'{self.timestamp}' WHERE family_id={family_id} and participant_id={participant_id} ;",
+                self.prod_connection,
+            )
+            self.assertEqual(entry, prod_random_results[0])
+
+    def test_random_rows_participant_phenotypes(self):
+        """ Test 10 random rows in the participant_phenotypes table """
+        restored_random_results = get_results(
+            f'SELECT * FROM participant_phenotypes ORDER BY RAND() LIMIT 10',
+            self.local_connection,
+        )
+        for entry in restored_random_results:
+            participant_id = entry[0]
+            hpo_term = entry[1]
+
+            prod_random_results = get_results(
+                f"SELECT * FROM participant_phenotypes FOR SYSTEM_TIME AS OF TIMESTAMP'{self.timestamp}' WHERE participant_id={participant_id} and hpo_term={hpo_term} ;",
+                self.prod_connection,
+            )
+            self.assertEqual(entry, prod_random_results[0])
+
+    @classmethod
+    def tearDownClass(cls):
+        """ Delete test database following testing """
+        subprocess.run(['rm', '-r', LOCAL_BACKUP_FOLDER], check=True)
+        subprocess.run(['sudo', 'rm', '-r', '/var/lib/mysql'], check=True)
+
+
+def pull_latest_backup():
+    """ Assumes no changes to metadata since initial upload """
+    storage_client = storage.Client()
+    blobs = [(blob, blob.updated) for blob in storage_client.list_blobs(BACKUP_BUCKET)]
+    latest = sorted(blobs, key=lambda tup: tup[1])[-1][0]
+    full_path = os.path.dirname(latest.name)
+    backup_folder = os.path.dirname(full_path)
+    return backup_folder
+
+
+def get_timestamp(folder: str):
+    """Returns timestamp in format YYYY-MM-DD HH:MM:SS """
+    timestamp = f'{folder[13:17]}-{folder[10:12]}-{folder[7:9]} {folder[18:20]}:{folder[21:23]}:{folder[24:26]}'
+    return timestamp
+
+
+def get_databases(connection: ConnectionDetails):
+    """ Returns a list of the databases """
+    conn = mysql.connector.connect(user=connection.user, host=connection.address)
+    cursor = conn.cursor()
+    cursor.execute('show databases')
+    databases = list(cursor)
+    return databases
+
+
+def get_results(query: str, connection: ConnectionDetails):
+    """ Returns the results from a provided query at a given connection """
+    conn = mysql.connector.connect(user=connection.user, host=connection.address)
```

Comment:
Can we create the connection once (rather than pass the credentials), and reuse it (by creating a new cursor) for each request?

A `mysql.connector` allows you to specify the `database` to connect to up front (to avoid the separate USE statement): https://dev.mysql.com/doc/connector-python/en/connector-python-connectargs.html#:~:text=Table%C2%A07.1%C2%A0Connection%20Arguments%20for%20Connector/Python